### ISSUE TITLE
`.readAllBytes()` on `InputStream` causes Runtime exceptions

### ISSUE THREAD

**Issue by TonioGela** [scala-native/scala-native#2973]
Hello! This is my first bug on Scala native, and I want to point out that I'm a total newbie with native (while not at all with Scala on JVM), so I might be missing something extremely silly. OTOH it's @lolgab that suggested me to open the issue 😇.

Using `scala-cli` I noticed a strange behaviour. Having this folder structure

```md
.
├── foo
│   └── foo <- contains "foo"
└── foo.scala
```

and this `foo.scala` file:

```scala
//> using scala "2.13.10"
//> using platform "scala-native"
//> using resourceDir "./foo"

import java.io.InputStream

object Main extends App {
    val array: Array[Byte] = getClass().getResourceAsStream("foo").readAllBytes()
    println(new String(array))
}
```

a mere `scala-cli run foo.scala` causes at runtime this Exception being thrown

```
scala.scalanative.runtime.UndefinedBehaviorError
        at java.lang.StackTrace$.currentStackTrace(Unknown Source)
        at java.lang.Throwable.fillInStackTrace(Unknown Source)
        at scala.scalanative.runtime.package$.throwUndefined(Unknown Source)
        at Main$.delayedEndpoint$Main$1(Unknown Source)
        at Main$delayedInit$body.apply(Unknown Source)
        at scala.Function0.apply$mcV$sp(Unknown Source)
        at scala.runtime.AbstractFunction0.apply$mcV$sp(Unknown Source)
        at scala.App.$anonfun$main$1(Unknown Source)
        at scala.App$$Lambda$1.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(Unknown Source)
        at scala.collection.AbstractIterable.foreach(Unknown Source)
        at scala.App.main(Unknown Source)
        at Main.main(Unknown Source)
        at <none>.main(Unknown Source)
```

while commenting out `//> using platform "scala-native"` and thus using JVM simply prints out the string `foo`

Suspecting that `readAllBytes` could have been the culprit I've tried reimplementing the same logic using a Java pre 9 way (If I'm not wrong `readAllBytes` was introduced in Java 9):

```scala
//> using scala "2.13.10"
//> using platform "scala-native"
//> using resourceDir "./foo"

import java.io.InputStream

object Main extends App {
    val stream: InputStream = getClass().getResourceAsStream("foo")
    val array: Array[Byte] = Array.ofDim[Byte](10)
    stream.read(array, 0, 9)
    println(new String(array))
}
```

and surprisingly this method works, so I don't think it might be related to resources reading but to that method being called.

Hope it helps!

**Comment by armanbilge** [comment]
Thanks for the report. Indeed, `readAllBytes()` has not been implemented for `InputStream` in Scala Native's Javalib.

https://github.com/scala-native/scala-native/blob/a3df752e54173073ce88c758b8bd6868e9a123cb/javalib/src/main/scala/java/io/InputStream.scala

A PR to add it would be very welcome 😃 

Separately, it's strange that this manifests as a runtime error instead of a linking error.

**Comment by TonioGela** [comment]
> A PR to add it would be very welcome 😃

You don't mean by me right ? D:

**Comment by ekrich** [comment]
I have not experimented with Scala CLI but I think you would need to have `foo` in a `resources` directory - that is a Java thing. You would also have to set the Config setting for embedding resources to `true`. That may be why you get the runtime error. I would think you would get a linking error but can't remember if you need link stubs `true` as well. 

Looks like the list here is incomplete - https://scala-native.org/en/stable/user/sbt.html#sbt-settings-and-tasks

**Comment by armanbilge** [comment]
@ekrich sclaa-cli works differently. That's what this configures.
```
//> using resourceDir "./foo"
```

**Comment by ekrich** [comment]
So it knows if you add a resource directory you want embedded resources setting on I suppose. I think you would need a `scala-native` dir for glue `.c` code unless they really hacked things.

**Comment by TonioGela** [comment]
> Thanks for the report. Indeed, `readAllBytes()` has not been implemented for `InputStream` in Scala Native's Javalib.
> 
> https://github.com/scala-native/scala-native/blob/a3df752e54173073ce88c758b8bd6868e9a123cb/javalib/src/main/scala/java/io/InputStream.scala
> 
> A PR to add it would be very welcome 😃
> 
> Separately, it's strange that this manifests as a runtime error instead of a linking error.

Btw reading at the `InputStream` class, a super silly implementation (that it's not very much native probably) could be

```scala
def readAllBytes(): Array[Byte] = {
   val buffer: Array[Byte] = Array.ofDim[Byte](Int.MaxValue)
   val readBytes = read(buffer)
   buffer.take(readBytes)
}
```

but I don't know if in Scala Native there are memory constraints, if an Array of `Int.MaxValue` size will break something, how the garbage collectors work, if the implementation is too expensive in terms of memory and performance. I'm too used to JVM garbage collectors and to immutability 😇

**Comment by ekrich** [comment]
Better to take general discussion to Discord or to an issue about `readAllBytes`. Feel free to make issues.

**Comment by armanbilge** [comment]
> to an issue about `readAllBytes`.

Isn't that what _this_ issue is about? 😉

**Comment by TonioGela** [comment]
Another chance, discussed with @armanbilge is to implement the `transferTo` method on InputStream and use it to implement `readAllBytes` using a `ByteArrayOutputStream`

The implementation of java's `transferTo` (where `DEFAULT_BUFFER_SIZE = 8192`) is 
```java
public long transferTo(OutputStream out) throws IOException {
    Objects.requireNonNull(out, "out");
    long transferred = 0;
    byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
    int read;
    while ((read = this.read(buffer, 0, DEFAULT_BUFFER_SIZE)) >= 0) {
        out.write(buffer, 0, read);
        transferred += read;
    }
    return transferred;
}
```

a logic that can be copied almost identically, to be then able to implement `readAllBytes` like so:
```scala
def readAllBytes(): Array[Byte] =   {
    val bos: ByteArrayOutputStream = new ByteArrayOutputStream()
    transferTo(bos)
    val array: Array[Byte] = bos.toByteArray()
    bos.close()
    array
}
```

A possible implementation of `transferTo` instead, might look like this
```scala
def transferTo(out: OutputStream): Long = {
    var transferred: Long = 0
    val buffer: Array[Byte] = Array.ofDim[Byte](DEFAULT_BUFFER_SIZE)
    var read: Int = 0

    do {
        read = this.read(buffer,0,DEFAULT_BUFFER_SIZE)
        out.write(buffer, 0, read)
        transferred += read
    } while (read >= 0)

    transferred
}
```

What do you think?

**Comment by ekrich** [comment]
I think the first is better because do while is dropped in Scala 3 -
https://dotty.epfl.ch/docs/reference/dropped-features/do-while.html

**Comment by LeeTibbert** [comment]
@TonioGella First & most importantly, thank you for the base note reporting the problem, your careful explanation, and
reproducers.

I spent some time studying this before the above entries, so I will try to not duplicate what others have said.
Although I think Arman hit it spot-on when he said that it is surprising indeed that this is not a link error.

The SN javalib [docs](https://scala-native.org/en/latest/lib/javalib.html) are, to my reading, exceedingly quiet about 
which version of Javalib SN supports.  

I have always been taught that SN links against some form of Java 8 (classpath mumble, mumble).  
I can never quite tell when a Java > 8 method may or may not  make it past link.  
Well, the linker usually tells me when I fail, but by then I have invested _way_ too much time.  
So I avoid anything above Java 8.  One person's survival strategy.

Can anyone give an authoritative description of  the current rule for which version of Javalib SN supports?  
If yes, I will create an Issue & PR to make that document clearer.  Indeed, a corresponding section in the Contributor's Guide
is also probably in order.  

That way, we can use TonioGella's experience to advance the ecosystem.

**Comment by armanbilge** [comment]
> Can anyone give an authoritative description of the current rule for which version of Javalib SN supports?

Yes. Scala Native in theory supports the latest JDK (or at least the latest LTS, which is 17). Of course, it's up to us to implement that support. So in practice what this means is we will not reject a PR for adding a Javalib method that is "too new", that would be silly :)

---

> I have always been taught that SN links against some form of Java 8 (classpath mumble, mumble).

Scala Native compiles against whatever JDK is on the classpath. So if you are building a Scala Native project with JDK 17, then you are compiling against JDK 17.

Scala Native links against the Javalib, as defined in this repository. So long as a JDK method is implemented in the Javalib here, then it will link.

---

> I can never quite tell when a Java > 8 method may or may not make it past link.

Right, it depends whether somebody has implemented in the Javalib or not :)

**Comment by LeeTibbert** [comment]
Arman,   Thank you for the patient explanation.  If I ever get free of javalib Process, I will condense that  for a doc PR.
I do hope for a more 'official' signoff before investing time in a PR.

I think the traditional issue has been that only java 8 things were implemented in javalib because one could not
tell the java version on the linking system, beyond Java 8 being the lowest supported versions.  

If a Java 17 method were implemented, if used, it would cause a link error when an application developer
linked against Java 8.  Probably no worse or unpredictable than today's behavior.


I think the SN build used to flat out fail with anything not Java 8.  Times change.

**Comment by armanbilge** [comment]
> If a Java 17 method were implemented, if used, it would cause a link error when an application developer
linked against Java 8.

We already have JDK 9+ methods implemented in Javalib :) the reason they don't cause problems for developers using JDK 8, is because they simply cannot compile code using JDK 9+ methods because they only have JDK 8 on the classpath.

So to clarify: if we could wave a magic wand and implement all of Javalib up to JDK 17, this would not create any "unpredictability" for any developers. They would be able to use exactly the JDK on their classpath.

**Comment by armanbilge** [comment]
> when an application developer linked against Java 8.

Sorry, one more point of clarification :)

Scala Native programs never _link_ against a particular version of the JDK. They _always_ link against the Javalib that is implemented in this repository. That is because linking works with NIR, and the Javalib implemented here is the one and only NIR artifact for Javalib.

_However_, Scala Native programs may _compile_ against different versions of the JDK. This is because compiling works against class files, and there are many different classfiles versions of the JDK. This depends on the JDK version you have on your classpath.

**Comment by LeeTibbert** [comment]
Arman.  Thank you again for explaining.   Makes sense when you walk me through it.

**Comment by LeeTibbert** [comment]
Because I was/am blocked on my main lines of endeavor, I took a quick look at implementing Streams `readAllBytes`.

It is not at the deep end of the pool, but certainly past the 2 metre level. i.e. the water is above nose level.

Implementing `readNBytes` is certainly feasible in economic time. 

`readAllBytes` falls into the "unknown & potentially effectively infinite buffer size" problem of
StringBuffer and all its issues of growth progression (i.e. the oft used  'times 1.5' is simply daft, not
to put too fine a point on it.) and wasted memory.    A Small Matter of Probraming but  bigger than a breadbox.

Using `readNBytes` requires having some estimate of the maximum size of the file and of 
the amount of memory they are willing to waste if the file is small.  I can see a usage where
someone does a readNBytes of 1024 for files known to be 'small', with a test for 'retry required'.


So, anybody have a use in the next six months for Stream `readNBytes`. I think there are
two variants.  

If this is a 'Java 9' completeness issue and not an active use issue, I think there are 
Java 8 completeness (and correctness) issues ahead of it.

**Comment by LeeTibbert** [comment]
@TonioGela @arman @ekrich 

I am looking at this issue as respite from an even harder one.

A question please, about what would be useful.

`readAllBytes()` & `readNBytes(buf, offset, len)` were introduced in Java 9.
`readNBytes(n)` was introduced in Java 11.  
Java 11 is the "next" long term supported version after Java 8. 

The natural & documented (Java 18ish) way to implement `readAllBytes()` is
in terms of the Java 11 `readNBytes(n)` (where n is given as MaxInt).

If I implement for Java 11, would that be useful to you TonioGela or
if be like Tantalus, existing but just out of reach?

If I am investing effort, I would like to do it for the longer-mid-term but I also
would like to see it useful soon (0.5.0, 0.4.?).

Thank you for your thoughts.

Lee

**Comment by ekrich** [comment]
This sounds fine and seems quite doable - any additions are certainly ok for back port as well.

**Comment by LeeTibbert** [comment]
@TonioGela @lolgab    PR #2031 has been merged. It is a candidate for both 0.5.0 & the 0.4.n tracks.

Thank you for providing both the information & the  impetus to getting that bit done.

The `readNBytes` part looks pretty useful.  I do not know how many times in my life 
I have written "read() may not read everything requested, so this need re-try code".
Not that re-try is already done and in one place.

If you experience any problems or have concerns with this implementation, please 
let me know.  Sorry the solution is weeks late and a Euro short.

I found a way around the "growable array" difficulty. It is not a particularly
efficient algorithm (too many copies) but it is done and not in-waiting.
